client/components/token-field/token-input.jsx

```diff
@@ -1,56 +1,63 @@
+/** @format */
 /**
  * External dependencies
- *
- * @format
  */
-
 import PropTypes from 'prop-types';
 import React from 'react';
 import PureRenderMixin from 'react-pure-render/mixin';
+import { omit, noop } from 'lodash';
 
 const TokenInput = React.createClass( {
 	propTypes: {
+		disabled: PropTypes.bool,
+		hasFocus: PropTypes.bool,
 		onChange: PropTypes.func,
 		onBlur: PropTypes.func,
-		value: PropTypes.string,
 		placeholder: PropTypes.string,
-		disabled: PropTypes.bool,
+		value: PropTypes.string,
 	},
 
 	getDefaultProps: function() {
 		return {
-			onChange: function() {},
-			onBlur: function() {},
-			value: '',
 			disabled: false,
+			hasFocus: false,
+			onChange: noop,
+			onBlur: noop,
 			placeholder: '',
+			value: '',
 		};
 	},
 
 	mixins: [ PureRenderMixin ],
 
+	componentDidUpdate: function() {
+		if ( this.props.hasFocus ) {
+			this.textInput.focus();
+		}
+	},
+
 	render: function() {
-		const props = { ...this.props, onChange: this._onChange };
-		const { value, placeholder } = props;
+		const { placeholder, value } = this.props;
 		const size =
 			( ( value.length === 0 && placeholder && placeholder.length ) || value.length ) + 1;
 
 		return (
-			<input ref="input" type="text" { ...props } size={ size } className="token-field__input" />
+			<input
+				className="token-field__input"
+				onChange={ this.onChange }
+				ref={ this.setTextInput }
+				size={ size }
+				type="text"
+				{ ...omit( this.props, [ 'hasFocus', 'onChange' ] ) }
```

Comment:
❤️  Much better

However, I'm trying to reason about why things were done the way they were, which was pretty strange. I suspect it has to do with what `props` we allow to be overwritten with the infamous `{...this.props}` AKA _whatever anybody may have stuck in here_.

## Before

Overwritable:
* `ref`
* `type`

Safe:
* `size`
* `className`
* `onChange`

## After

Overwritable:
* `ref`
* `type`
* `size`
* `className`

Safe:
* `onChange` (via `omit`)

Reply:
Yeah, this still isn't great. As always, more explicit props would be nicer. I guess I didn't want to break consumers that were passing any obscure props here, but then again, this is only used by https://github.com/Automattic/wp-calypso/blob/13e18c519375819c89d11145e08efc0ff3ef2980/client/components/token-field/index.jsx#L194 anyway. I'll give this another look...

Reply:
Hmm, I think I'll leave as-is after all for now. Just don't want to risk to introduce any breakage because I miss a little prop now that you've reviewed and tested this one. Also, re-doing props should be a fairly easy and mechanical task we can tackle later.